Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIPermissionsCocoa.mm

```diff
@@ -123,34 +141,33 @@
 
 bool WebExtensionAPIPermissions::verifyRequestedPermissions(HashSet<String>& permissions, HashSet<Ref<WebExtensionMatchPattern>>& matchPatterns, NSString *callingAPIName, NSString **outExceptionString)
 {
-    WebExtension extension = WebExtension(extensionContext().manifest());
-    HashSet<String> allowedPermissions = extension.requestedPermissions();
+    auto extension = WebExtension::create(extensionContext().manifest(), @{ });
+    HashSet<String> allowedPermissions = extension->requestedPermissions();
 
-    if ([callingAPIName isEqualToString:@"remove()"]) {
+    if ([callingAPIName isEqualToString:@"permissions.remove()"]) {
         if (bool requestingToRemoveFunctionalPermissions = permissions.size() && permissions.isSubset(allowedPermissions)) {
-            *outExceptionString = @"Required permissions cannot be removed.";
+            *outExceptionString = @"You cannot remove required permissions.";
             return false;
         }
     }
 
-    allowedPermissions.add(extension.optionalPermissions().begin(), extension.optionalPermissions().end());
+    allowedPermissions.add(extension->optionalPermissions().begin(), extension->optionalPermissions().end());
 
-    bool requestingPermissionsNotDefinedInManifest = permissions.size() && !permissions.isSubset(allowedPermissions);
+    bool requestingPermissionsNotDeclaredInManifest = permissions.size() && !permissions.isSubset(allowedPermissions);
+    if (requestingPermissionsNotDeclaredInManifest) {
+        // Don't report any errors for remove if the permission isn't declared in the manifest, just silently return false.
+        *outExceptionString = [callingAPIName isEqualToString:@"permissions.request()"] ? @"Only permissions specified in the manifest may be requested." : nil;
+        return false;
+    }
 
-    WebExtension::MatchPatternSet allowedPatterns = extension.requestedPermissionMatchPatterns();
+    WebExtension::MatchPatternSet allowedPatterns = extension->requestedPermissionMatchPatterns();
     for (auto& requestedPattern : matchPatterns) {
-        bool requestingAllowedOrigin = false;
         for (auto& allowedPattern : allowedPatterns) {
-            if (requestedPattern->matchesPattern(allowedPattern, { WebExtensionMatchPattern::Options::IgnorePaths })) {
-                requestingAllowedOrigin = true;
-                break;
+            if (bool requestingDisallowedOrigin = !requestedPattern->matchesPattern(allowedPattern, { WebExtensionMatchPattern::Options::IgnorePaths })) {
```

Comment:
I think this method needs a bit more work. I believe I should be adding the optional match patterns here as well and not return an error message if the requested pattern isn't in the manifest. will confirm in the morning when I re-test this.